### PR TITLE
fix #1558 Add value formatter(s) to StepVerifierOptions

### LINKED ISSUE #1558: Improve diagnosability for data signal mismatches in StepVerifier with a non-helpful toString() method
Currently, if an expectation for a data signal fails (`StepVerifier.create(…).expectNext(foo)`), the error message contains a `toString()` representation of the expected and the actual data item. This works very well if the class of the emitted item reports sufficient information to diagnose a mismatch (e.g. `@Data` classes, `String`, most `java.lang` types).

Some types (such as Java arrays, `ByteBuffer`) report information in their `toString()` method that aren't useful for further diagnosis:

```
expectation "expectNext(java.nio.HeapByteBuffer[pos=0 lim=7 cap=7])" failed (expected value: java.nio.HeapByteBuffer[pos=0 lim=7 cap=7]; actual value: java.nio.HeapByteBuffer[pos=0 lim=7 cap=7])
```

It would be helpful to provide more insight to the actual payload, and ideally `StepVerifier` ships with formatters (`ToStringFunction`?) for common and well-known Java types that suffer from an informative `toString()` method.

I'm not sure that it makes a lot of sense to provide a user-configurable API because reactive code comes with a certain complexity and so does test code. We could continue writing test code `mono.as(StepVerifier::as).withOptions(…)` but I'm a bit torn regarding value vs. complexity.

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

**Comment by simonbasle** [comment]
mainly missing test coverage. @mp911de what do you think?

**Comment by mp911de** [comment]
It's evolving towards a great direction. I would broaden the value formatter predicate to `Predicate<Object>` instead of `Class<?>`. This would allow generic handling for e.g. array types and format an array properly with a single generic formatter instead of registering formatters for each array-type.

Another thing: Would it make sense to ship with built-in formatters for arrays and `ByteBuffer`/`ByteBuf`?

**Comment by simonbasle** [comment]
Default formatters for arrays might be interesting. `ByteBuf` isn't possible as there is no dependency to netty in core

**Comment by simonbasle** [comment]
@mp911de reworked the PR in a simplified direction where each `StepVerifier` would need a dedicated `StepVerifierOptions` configured with a single formatter, but one tailored for the type `<T>` of `Flux<T>` under test. The caveat is that the `Function<Object, String>` could be applied to stuff other than what goes through the `Flux`, as it is used to format messages in logger-with-varargs style.

I added utility methods to produce such formatters based on a `Class`, a `Predicate` or on container classes like arrays, iterables and `Signal<T>` (where `T` is the class for which we want to override `toString()`).

File: reactor-test/src/main/java/reactor/test/ValueFormatters.java [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+public final class ValueFormatters {
+
+	private ValueFormatters() {}
+
+	/**
+	 * A generic {@link Object} to {@link String} conversion {@link Function} which is
+	 * also a {@link Predicate}, and which only applies a custom conversion to targets that
+	 * match said {@link Predicate}. Other targets are converted using {@link String#valueOf(Object)}.
+	 */
+	interface ToStringConverter extends Predicate<Object>, Function<Object, String> {}
```
Comment (mp911de) [inline reactor-test/src/main/java/reactor/test/ValueFormatters.java]: `ToStringConverter` needs to be `public`

Comment (simonbasle) [reply reactor-test/src/main/java/reactor/test/ValueFormatters.java]: ah good catch, both interfaces needs to be `public` yeah

**Comment by smaldini** [comment]
I wonder if we should have an extension for formatters so netty actually participates in ?

**Comment by simonbasle** [comment]
@smaldini maybe at least an extension point, like a static method to add default extractors?
`ByteBuf` can then be covered by an extractor that uses eg.:
 - `byteBuf.getClass().getSimpleName() + "{` as the _prefix_
 - `Stream.of(byteBuf.toString(Utf8))` as the _extracted content Stream_
 - `"}"` as the _suffix_

**Pull request merged**